SYCL/GroupAlgorithm/barrier.cpp

```diff
@@ -0,0 +1,158 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_80
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+
+// REQUIRES: cuda
+
+#include "CL/sycl.hpp"
+#include <iostream>
+#include <vector>
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental::cuda;
+
+void basic() {
+  queue q{};
+  int N = 64;
+  std::vector<int> data(N);
+  for (int i = 0; i < N; i++) {
+    data[i] = i;
+  }
+  {
+    buffer<int> buf(data.data(), N);
+
+    q.submit([&](handler &cgh) {
+      auto acc = buf.get_access<access::mode::read_write>(cgh);
+      accessor<int, 1, access::mode::read_write, access::target::local> loc(
+          N, cgh);
+      accessor<barrier, 1, access::mode::read_write, access::target::local>
+          loc_barrier(2, cgh);
+      cgh.parallel_for(nd_range<1>(N, N), [=](nd_item<1> item) {
+        size_t idx = item.get_local_linear_id();
+        loc[idx] = acc[idx];
+        if (idx <= 2) {
+          loc_barrier[idx].initialize(N);
+        }
+        item.barrier(access::fence_space::local_space);
+        for (int i = 0; i < N; i++) {
+          int val = loc[idx];
+          barrier::arrival_token arr = loc_barrier[0].arrive();
+          val += 1;
+          int dst_idx = (idx + 1) % N;
+          loc_barrier[0].wait(arr);
+          loc[dst_idx] = val;
+          loc_barrier[1].wait(loc_barrier[1].arrive());
+        }
+        acc[idx] = loc[idx];
+      });
+    });
+  }
+  for (int i = 0; i < N; i++) {
+    assert(data[i] == i + N);
+  }
+}
+
+void interface() {
+  queue q{};
+  int N = 64;
+  std::vector<int> data(N, -1);
+  std::vector<int> test1(N, -1);
+  std::vector<int> test2(N, -1);
+  for (int i = 0; i < N; i++) {
+    data[i] = i;
+  }
+  {
+    buffer<int> data_buf(data.data(), N);
+    buffer<int> test1_buf(test1.data(), N);
+    buffer<int> test2_buf(test2.data(), N);
+
+    q.submit([&](handler &cgh) {
+      auto data_acc = data_buf.get_access<access::mode::read_write>(cgh);
+      auto test1_acc = test1_buf.get_access<access::mode::read_write>(cgh);
+      auto test2_acc = test2_buf.get_access<access::mode::read_write>(cgh);
+      accessor<int, 1, access::mode::read_write, access::target::local> loc(
+          N, cgh);
+      accessor<barrier, 1, access::mode::read_write, access::target::local>
+          loc_barrier(2, cgh);
+      stream s(1024 * 1024, 1024, cgh);

```

Comment:
Is this used anywhere?

Reply:
No, this was left in from debugging. Removed.